runtime/lua/vim/filetype.lua

```diff
@@ -1480,8 +1480,18 @@ end
 local pattern_sorted = sort_by_priority(pattern)
 
 ---@private
-local function normalize_path(path)
-  return (path:gsub("\\", "/"):gsub("^~", vim.env.HOME))
+local function normalize_path(path, as_pattern)
+  local normal = path:gsub("\\", '/')
+  if normal:find('^~') then
+    if as_pattern then
+      -- Escape Lua's metacharacters when $HOME is used in a pattern.
+      -- The rest of path should already be properly escaped.
+      normal = vim.env.HOME:gsub('[-^$()%%.%[%]+?]', '%%%0') .. normal:sub(2)
```

Comment:
is `vim.pesc` useful here? https://github.com/neovim/neovim/blob/b470ff0091bbcce4f59fe6ef27d745d49a29c278/runtime/lua/vim/shared.lua#L507

Reply:
Yeah, that would have been useful to notice. :)

Reply:
There will be more than one opportunity to refactor `filetype.lua` (e.g., as part of #18241) ;)